prereise/gather/griddata/hifld/data_process/transmission.py

```diff
@@ -27,6 +30,161 @@ def check_for_location_conflicts(substations):
+        )
+
+    print(
+        f"initial number of lines: {lines.shape[0]}\n"
+        f"zero distance lines after rounding: "
+        f"{lines.shape[0] - line2sub['FROM'].compare(line2sub['TO']).shape[0]}\n"
+    )
+
+    from_sub_missing = line2sub["FROM_SUB"].isnull().values
+    to_sub_missing = line2sub["TO_SUB"].isnull().values
+
+    unique_endpoint_coord = set(
+        line2sub["FROM"].unique().tolist() + line2sub["TO"].unique().tolist()
+    )
+
+    n_two_sub_missing = sum(from_sub_missing & to_sub_missing)
+    n_one_sub_missing = sum(from_sub_missing ^ to_sub_missing)
+    n_unique_orphan_endpoint = len(unique_endpoint_coord - set(subcoord2subid))
+    n_unconnected_sub = len(set(subcoord2subid) - unique_endpoint_coord)
+
+    print(
+        f"there are {n_two_sub_missing} lines with two substations missing\n"
+        f"there are {n_one_sub_missing} lines with one substation missing\n"
+        f"there are {n_unique_orphan_endpoint} unique orphan endpoint\n"
+        f"there are {n_unconnected_sub} substations unconnected\n"
+    )
+
+    return line2sub
+
+
+def assign_substations_to_lines(substations, lines, line2sub):
+    """Locate closest substation of unconnected lines and update information
+    information of lines. Note that the ``line2sub`` data frame is modified in-place.
+
+    :param pandas.DataFrame substations: data frame of substations.
+    :param pandas.DataFrame lines: data frame of all lines.
+    :param pandas.DataFrame line2sub: indices: table mapping lines to substations as
+        returned by :func:`map_lines_to_substations_using_coords`.
+    :return: (*pandas.DataFrame*) -- original lines data frame with updated *SUB_1*
+        and *SUB_2* columns, new *SUB_1_ID*, *SUB_2_ID* along with *OTHER_SUB_1_ID*
+        and *OTHER_SUB_2_ID* when multiple location are associated to the same endpoint.
+    """
+    print("finding closest neighbor to unconnected lines")
+
+    # find closest neighbor of unconnected lines
+    subcoord2subid = substations.groupby(["LONGITUDE", "LATITUDE"]).groups
+    subcoord = list(subcoord2subid)
+    missing = line2sub.query("FROM_SUB.isnull() or TO_SUB.isnull()").index
+    endpoint2neighbor = {}
+    for i in tqdm(missing, total=len(missing)):
+        for endpoint in ["FROM", "TO"]:
+            if isinstance(line2sub.loc[i, f"{endpoint}_SUB"], float):
+                if line2sub.loc[i, endpoint] in endpoint2neighbor:
+                    line2sub.loc[i, f"{endpoint}_SUB"] = endpoint2neighbor[
+                        line2sub.loc[i, endpoint]
+                    ]
+                else:
+                    subid = subcoord2subid[
+                        subcoord[
+                            find_closest_neighbor(line2sub.loc[i, endpoint], subcoord)
+                        ]
+                    ]
+                    line2sub.loc[i, f"{endpoint}_SUB"] = list(subid)
+                    endpoint2neighbor[line2sub.loc[i, endpoint]] = list(subid)
```

Comment:
I took a stab at reimplementing this in a way that seemed simpler/clearer to me, since it seems we can in order:
1. Find substation(s) closest to all endpoints with no substation listed (`endpoint2neighbor`)
2. Look up the closest substations for each missing endpoint (new `filled_subs` variable)
3. Update the `line2sub` dataframe with the information from (2)
```python
end_sub = {"FROM": "FROM_SUB", "TO": "TO_SUB"}
missing_points = set().union(
    *[set(line2sub[e].loc[line2sub[e_sub].isna()]) for e, e_sub in end_sub.items()]
)
endpoint2neighbor = {
    point: subcoord2subid[subcoord[find_closest_neighbor(point, subcoord)]]
    for point in tqdm(missing_points, total=len(missing_points))
}
filled_subs = [
    line2sub[e].loc[line2sub[e_sub].isna()].map(endpoint2neighbor).map(list)
    for e, e_sub in end_sub.items()
]
line2sub.update(pd.concat(filled_subs, axis=1).rename(columns=end_sub))
```

It saves a few lines, but not as many as I had originally hoped.

Reply:
Works for me. We have the same kind of performance (2:40s for rounding=2)

Reply:
Done